### PR TITLE
Changes for Embarcadero C++ clang-based compilers, targeting Boost 1.74

### PR DESCRIPTION
**Pull request by eldiener**

### CODE REVIEW

**Comment by eldiener** [comment]
A code coverage check failed with:

```
+lcov --version

lcov: LCOV version 1.14-6-g40580cd

+lcov --gcov-tool=gcov-7 --rc lcov_branch_coverage=1 --base-directory /home/travis/build/boostorg/boost-root/libs/program_options --directory /home/travis/build/boostorg/boost-root --capture --output-file all.info

Capturing coverage data from /home/travis/build/boostorg/boost-root

Can't locate PerlIO/gzip.pm in @INC (you may need to install the PerlIO::gzip module) (@INC contains: /etc/perl /usr/local/lib/perl/5.18.2 /usr/local/share/perl/5.18.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.18 /usr/share/perl/5.18 /usr/local/lib/site_perl .) at /usr/local/bin/geninfo line 62.

BEGIN failed--compilation aborted at /usr/local/bin/geninfo line 62.
```

This has nothing to with this PR. Can this please be merged ?

**Comment by pdimov** [comment]
The change from `abort` to `std::abort` seems wrong; `std::abort` is defined in `<ctsdlib>`, whereas `<stdlib.h>`, which is being included, defines `abort`.

File: test/winmain.cpp [hunk 1]
```diff
@@ -10,12 +10,10 @@
 #include <iostream>
 #include <stdlib.h>
 
-using namespace std;
-
 #include <boost/program_options/parsers.hpp>
 using namespace boost::program_options;
 
-void check_equal(const std::vector<string>& actual, char **expected, int n)
+void check_equal(const std::vector<std::string>& actual, char **expected, int n)
```
Comment (vprus) [inline test/winmain.cpp:16]: Is this change necessary to make borland work?

File: test/winmain.cpp [hunk 2]
```diff
@@ -40,7 +38,7 @@ void test_winmain()
 #define C ,
 #define TEST(input, expected) \
     char* BOOST_PP_CAT(e, __LINE__)[] = expected;\
-    vector<string> BOOST_PP_CAT(v, __LINE__) = split_winmain(input);\
+    std::vector<std::string> BOOST_PP_CAT(v, __LINE__) = split_winmain(input);\
```
Comment (vprus) [inline test/winmain.cpp:41]: Likewise.

**Review by vprus — CHANGES_REQUESTED** [review]
Thanks for the patch! The main chance seems straightforward, but looks some some unrelated cleanups got added?

File: test/winmain.cpp (hunk 2 above)
Comment (eldiener) [reply test/winmain.cpp:41]: Please, this is C++. You really want to have 'using namespace std;', include C++ and C header files, and leave off the std:: namespaces everywhere you can ? All that together can lead to problems, and did wirth the Embarcadero C++ compiler. The code already produces warnings on the invocation of TEST, which I did not bother to change, trying to keep my changes to a minimum. How am I hurting your code by providing the std:: namespace excplicitly ? It is not breaking anything and is certainly avoiding any other potential problems.

Comment (vprus) [reply test/winmain.cpp:41]: All this is valid C++ that no other compiler had problem with for a decade. I think it's reasonable to ask what specific problem Borland had here

Comment (eldiener) [reply test/winmain.cpp:41]: The specific problem Embarcadero had is that there is a `std::byte` ( https://en.cppreference.com/w/cpp/types/byte ) and there is an unqualified `byte` in the Windows header file rpcndr.h. Having `using namespace std;` creates the clash. Before you ask why Embarcadero encounters this problem while other compilers do not, let me answer that I did not want to spend hours trying to figure this out rather than spend minutes fixing it with perfectly reasonable code which changes nothing as far as the logic goes. Really Vladimir, you know as well as I do that specifying `using namespace std;` is generally a poor choice for anything but the simplest code which #includes very little and using any Boost library's code is never that simple.

Comment (eldiener) [reply test/winmain.cpp]: I fixed the test/winmain.cpp in a better way without the previous changes I had made.

**Review by vprus — APPROVED** [review]
Thanks again for the patch, it looks good.
